src/WebJobs.Script/Host/IHostFunctionMetadataProvider.cs

```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Workers.Rpc;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    internal interface IHostFunctionMetadataProvider
```

Comment:
To clarify my understanding - does each metadata provider correspond to a different metadata type/class? 

If keeping separate interfaces, would it make sense to extend the specification in expectations to the methods in the interfaces too instead of just keeping the difference at the interface names? e.g. in IHostFunctionMetadataProvider, GetHostFunctionMetadataAsync instead of GetFunctionMetadataAsync?

Reply:
The interfaces exist only for unit testing - being able to provide mocks. Other than that, this is a very targeted scenario for the host vs worker function indexing logic, which necessitates slight deviations in the contract between the two. This prevents re-use of the `IFunctionMetadataProvider` for the worker logic. The host could use it sure - and I am not against that. The worker logic cannot use it (it needs a different return type). But overall, we have a targeted scenario here and not a general extensibility point. We are weighing the cost vs benefit of creating a contract which satisfies the scenario, but also does not betray the implementation details.

The multiple interfaces may look excessive, and they may honestly be just that. But there is not too much of a benefit to solving it differently. Although it may be worth making all of these very explicit. @soninaren - can you add comments explaining this? Or alternatively, you could even nest these interfaces within `DefaultFunctionMetadataProvider` to show that they are to be used **only** for that implementation.

But also, if we feel that we should not be introducing one-off interfaces like this as a matter of principal, then I also am not against that.

Reply:
Below is my proposed structuring which I feel will solve all of the concerns above.

1. IFunctionMetadataProvider (base interface)
        - Method - `GetMetadataAsync()`
2. IHostFunctionMetadataProvider : IFunctionMetadataProvider
        - No additional methods apart from `base.GetMetadataAsync()`
3. IWorkerFunctionMetadataProvider : IFunctionMetadataProvider
        - Additional Method - `IsFallbackNeededToHostIndexing()`

Reply:
@jviau that makes sense! Synced with @soninaren offline and also think it makes sense to keep them separate/that comments sound like a good idea!